### PR TITLE
feat: generate alignment pileups

### LINKED ISSUE #4: Optionally generate alignment pileups for selected regions
The following tool could be integrated in the `quant` workflow to visualize, e.g., isomiR or mutated miR coverage:
https://git.scicore.unibas.ch/zavolan_group/tools/ascii-alignment-pileup

### PR DESCRIPTION
**Pull request by deliaBlue**
This PR closes #4 .

The changes in this PR are:

- Add the script that generates the ASCII-style pileups
- Add the workflow that generates the pileups (`pileup.smk`)
- Add a new function in the `common.smk` to generate an empty BED file if not provided by the user
- Add the new workflow and its output files in the main Snakefile
- Add the pileups parameters in the `config_template.yaml`, `config_schema.json` and `test/config.yaml` files
- Add the BED files with the regions of interest as a test input file
- Document the new workflow in the main README, the configuration README and in the `pipeline_documentation.md` 
- Update the expected output and the rule graph

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -278,6 +284,10 @@ represents a sample library. Each read is counted towards all the annotated
 miRNA species it aligns to, with 1/n, where n is the number of genomic and/or
 transcriptomic loci that read aligns to.
 
+5. ASCII-style pileups of read alignments for the whole run, across libraries,
+and across subset of libraries - if specified - against the regions in the
+user-provided BED file, if any.
```
Comment (uniqueg) [inline README.md:289]: "for the whole run, across libraries and across subsets of libraries" is ambiguous/vague. Maybe better something like this:

**OPTIONAL** ASCII-style pileups of read alignments produced for individual libraries, combinations of libraries and/or all libraries of a given run.  The exact number and nature of outputs depends on the workflow inputs/parameters.

File: README.md [hunk 2]
```diff
@@ -322,7 +332,8 @@ that only the best alignments of each read are kept (smallest edit distance).
 Finally, alignments are intersected with the user-provided, pre-processed
 miRNA annotation file using [BEDTools][bedtools]. Counts are tabulated 
 separately for reads consistent with either miRNA precursors, mature miRNA
-and/or isomiRs.
+and/or isomiRs. Finally, if ASCII-style alignment pileups are generated across
+samples, for the whole run, and across conditions if specified.
```
Comment (uniqueg) [inline README.md:336]: Now you have "Finally" twice :upside_down_face:

Also, the new sentence is incomplete. If x, then ... what? Maybe rather something like:

Finally, ASCII-style alignment pileups are optionally generated for user-defined regions of interest.

Done. But remove the first "Finally" ;-)

File: config/README.md [hunk 3]
```diff
@@ -104,6 +104,11 @@ There are 4 files you must provide:
 > filtering), you can do that, but make sure the formats of any modified
 > resource files meet the formatting expectations outlined above!
 
+In addition, if you want to generate
+[ASCII-style alignment pileups][ascii-pileups], a **BED6** file following the
+[Ensembl format][ensembl-bed] with the regions of interest you want the pileups
+to be performed on.
+
```
Comment (uniqueg) [inline config/README.md:111]: Not sure why this is copied over from `README.md`? Is the entire section copied? Is this really necessary? Best to have only a single source of truth, otherwise we will forget updating both copies sooner or later if/when something changes (probably sooner)

File: config/config_schema.json [hunk 4]
```diff
@@ -9,6 +9,11 @@
             "type": "string",
             "description": "Path to the samples table."
         },
+        "bed_file":{
+            "type": "string",
+            "default": "",
+            "description": "Path to the genomic regions file to do the pileups on."
```
Comment (uniqueg) [inline config/config_schema.json:15]: Better: "generate the pileups for"

File: config/config_schema.json [hunk 5]
```diff
@@ -101,7 +111,16 @@
...
+            "default": ["isomir", "mirna", "pri-mir"],
+            "description": " miRNA speices to be quantified."
+        },
+        "lib_dict":{
+            "type": "object",
+            "additionalProperties":{
+                "type": "array",
+            },
+            "default": {},
+            "description": "Dictionary with a list of library names, as in the samples table, to perfrom the pileups on as value and a condition name as keys."
```
Comment (uniqueg) [inline config/config_schema.json:123]: Again, better something like: "to generate aggregate pileups for"

File: config/config_template.yaml [hunk 6]
```diff
@@ -63,4 +65,15 @@ nh: 100 # discard reads with more mappings than the indicated number
 # If 'isomir' and 'mirna' are both in the list, a single table with both types
 # is made.
 mir_list: ['isomir', 'mirna', 'pri-mir']
+
+#### ASCII-STYLE ALIGNMENT PILEUPS PARAMETERS ####
+
+# Dictionary with the list of library names to group when performing the
```
Comment (uniqueg) [inline config/config_template.yaml:71]: maybe better "aggregate" rather than "group" as it's more descriptive / less vague

File: config/config_template.yaml [hunk 7]
```diff
@@ -63,4 +65,15 @@ nh: 100 # discard reads with more mappings than the indicated number
...
+# pileups as values and the condition as keys. Library names must match the
+# ones in the samples' table `sample` column. The dictionary keys will be used
+# as the pileups output directory name.
```
Comment (uniqueg) [inline config/config_template.yaml:74]: "pileup's" rather than "pileups", I guess

File: pipeline_documentation.md [hunk 8]
```diff
@@ -87,6 +94,7 @@ on installation and usage please see [here](README.md).
 
 | Name | License | Tag line | More info |
 | --- | --- | --- | --- |
+| **ASCII-style alignment pileups** | [Apache 2.0][license-apache2] | _"Generates an ASCII-style pileup of read alignments in one or more BAM files against one or more regions specified in a BED file."_ | [code][code-ascii] |
```
Comment (uniqueg) [inline pipeline_documentation.md:97]: I think the description is a bit inaccurate:
- `Generates an`: I think the script can generate multiple pileups in one run, no? So should be plural.
- `against`: I think "for" is actually more accurate :)
- `regions specified in a BED file`: Maybe not so interesting here that it requires regions in a BED format. Rather I would just say "for one or more genomic regions."

File: scripts/ascii_alignment_pileup.R
Comment (uniqueg) [inline scripts/ascii_alignment_pileup.R:1]: Not reviewing that, obviously - trusting that this works :)

File: test/test_files/pileup_regions.bed
Comment (uniqueg) [inline test/test_files/pileup_regions.bed:1]: Did you link to that example somewhere when you describe these inputs or otherwise include an example? It's always easier than just saying something like **BED6** or link to an "Ensembl format" etc.

**Review by uniqueg — CHANGES_REQUESTED** [review]
Looks good. Few minor issues

File: config/README.md (hunk 3 above)
Comment (deliaBlue) [reply config/README.md:111]: It is not 100% copied, but part of it. This README is required by the Snakemake Catalogue.  Only a few things should be modified there - if necessary - in the future. What should I do with it?

File: pipeline_documentation.md (hunk 8 above)
Comment (deliaBlue) [reply pipeline_documentation.md:97]: Changing it, but just for you to know, this is exactly what's on GitLab as the tool's description.

File: test/test_files/pileup_regions.bed
Comment (deliaBlue) [reply test/test_files/pileup_regions.bed:1]: I can either add the example in main README, or create a gist and refer the user there.

File: README.md [hunk 9]
```diff
@@ -209,18 +209,24 @@ There are 4 files you must provide:
    resource][chrMap] provides such files for various organisms, and in the
    expected format.
 
+5. **OPTIONAL**: A **BED6** file with regions for which to produce
+   [ASCII-style alignment pileups][ascii-pileups] for. If not provided, no
+   pileups will be generated. Refer [here][bed-format] to see the expected
+   format.
```
Comment (uniqueg) [inline README.md:215]: * One `for` is enough here: _"for which to produce [...] ~for~"_
* Better: `See here for the expected format` rather than `Refer here to see the expected format`

File: README.md [hunk 10]
```diff
@@ -278,6 +284,11 @@ represents a sample library. Each read is counted towards all the annotated
 miRNA species it aligns to, with 1/n, where n is the number of genomic and/or
 transcriptomic loci that read aligns to.
 
+5. **OPTIONAL**. ASCII-style pileups of read alignments produced for individual
+libraries, combinations of libraries and/or all libraries of a given run. The
+exact number and nature of the outputs depends on the workflow
+inputs/parameters.
```
Comment (uniqueg) [inline README.md:290]: Perhaps this could be linked to the section where you describe this in more detail?

File: config/README.md [hunk 11]
```diff
@@ -100,6 +100,11 @@ There are 4 files you must provide:
    resource][chrMap] provides such files for various organisms, and in the
    expected format.
 
+5. **OPTIONAL**: A **BED6** file with regions for which to produce
+   [ASCII-style alignment pileups][ascii-pileups] for. If not provided, no
+   pileups will be generated. Refer [here][bed-format] to see the expected
+   format.
```
Comment (uniqueg) [inline config/README.md:106]: As above. But I'm wondering: Why do we have the exact same docs in two different places? :thinking:

File: config/config_schema.json [hunk 12]
```diff
@@ -101,7 +111,16 @@
...
+            "default": ["isomir", "mirna", "pri-mir"],
+            "description": " miRNA speices to be quantified."
+        },
+        "lib_dict":{
+            "type": "object",
+            "additionalProperties":{
+                "type": "array",
+            },
+            "default": {},
+            "description": "Dictionary with a list of library names, as in the samples table, to generate pileups for as value and a condition name as keys."
```
Comment (uniqueg) [inline config/config_schema.json:123]: I find this sentence very hard to parse. Maybe:

Dictionary of arbitrary condition names (keys) and library names to aggregate alignment pileups for (values; MUST correspond to names in samples table).

File: config/config_schema.json [hunk 13]
```diff
@@ -101,7 +111,16 @@
                 "type": "string",
                 "enum": ["isomir", "mirna", "pri-mir"]
             },
-            "default": ["isomir", "mirna", "pri-mir"]
+            "default": ["isomir", "mirna", "pri-mir"],
+            "description": " miRNA speices to be quantified."
```
Comment (uniqueg) [inline config/config_schema.json:115]: Remove empty space before "miRNA"

File: config/config_template.yaml [hunk 14]
```diff
@@ -63,4 +65,15 @@ nh: 100 # discard reads with more mappings than the indicated number
 # If 'isomir' and 'mirna' are both in the list, a single table with both types
 # is made.
 mir_list: ['isomir', 'mirna', 'pri-mir']
+
+#### ASCII-STYLE ALIGNMENT PILEUPS PARAMETERS ####
+
+# Dictionary with the list of library names to aggregate when performing the
+# pileups as values and the condition as keys. Library names must match the
+# ones in the samples' table `sample` column. The dictionary keys will be used
```
Comment (uniqueg) [inline config/config_template.yaml:73]: For consistency: samples table, not samples' table

File: config/config_template.yaml [hunk 15]
```diff
@@ -63,4 +65,15 @@ nh: 100 # discard reads with more mappings than the indicated number
...
+# as the pileup's output directory name.
```
Comment (uniqueg) [inline config/config_template.yaml:74]: Is it really the output _directory_? Will this not be part of the file basename, rather? Fine if not, just want to make sure this is accurate

**Review by uniqueg — CHANGES_REQUESTED** [review]
Getting there :)

File: config/README.md (hunk 11 above)
Comment (deliaBlue) [reply config/README.md:106]: As I mentioned in a comment above:



>It is not 100% copied, but part of it. This README is required by the 
>Snakemake Catalogue. Only a few things should be modified there - if 
>necessary - in the future. What should I do with it?
>
>(This README only contains how to/set up things to run the workflow, and 
>the expected inputs and outputs)

File: config/config_template.yaml (hunk 15 above)
Comment (deliaBlue) [reply config/config_template.yaml:74]: Yes, in the scenario in which the user uses this parameter, the output directory would be `RESULTS/PILEUPS/{key}/`. Should I reformulate the sentence?

File: README.md (hunk 10 above)
Comment (deliaBlue) [reply README.md:290]: Will it be enough to link it to the "pileups subworkflow" section in the `pipeline_documentation.md` or should I link it there and add a small explanation prior to the first rule of the workflow?

File: config/config_template.yaml (hunk 15 above)
Comment (uniqueg) [reply config/config_template.yaml:74]: No, no, that's fine :)

File: config/README.md (hunk 11 above)
Comment (uniqueg) [reply config/README.md:106]: Ah, sorry, didn't catch that. Fine then.

File: README.md (hunk 10 above)
Comment (uniqueg) [reply README.md:290]: Yes, that's totally fine.

**Review by uniqueg — APPROVED** [review]
Thanks :)
